### PR TITLE
Move caching tests from test_dispatcher to test_caching

### PR DESCRIPTION
**Pull request by gmarkall**
This moves the cache tests from `test_dispatcher.py` and `test_cfunc.py` into `test_caching.py` - this organises things a little better as `test_dispatcher.py` was huge, `test_caching.py` contained very few cache tests, and the cfunc cache tests were a small part of `test_cfunc.py`. Other tests that test caching are already in their own dedicated files, so these are left as-is except for updating imports.

@stuartarchibald and I did discuss refactoring to collapse the `BaseCacheTest` / `BaseCacheUseCasesTest` split, but now I see that both of these are used in different places I decided against trying to do that - instead, it seems to make the code more clearly-structured to just rename `BaseCacheUseCasesTest` to `DispatcherCacheUseCasesTest`.

### CODE REVIEW

**Comment by gmarkall** [comment]
gpuci run tests

(because you just never know!... :-) )

File: numba/tests/test_dispatcher.py [hunk 1]
```diff
@@ -1003,846 +949,6 @@ def pyfunc(A, i, j):
...
-        out, err = popen.communicate()
-        if popen.returncode != 0:
-            raise AssertionError(
-                "process failed with code %s: \n"
-                "stdout follows\n%s\n"
-                "stderr follows\n%s\n"
-                % (popen.returncode, out.decode(), err.decode()),
-            )
-
-    def check_module(self, mod):
```
Comment (stuartarchibald) [inline numba/tests/test_dispatcher.py:1090]: @gmarkall from OOB discussion, this was the function that we established did nothing/was never run?

File: numba/tests/test_caching.py [hunk 2]
```diff
@@ -133,3 +192,886 @@ def test(num=1000):
...
+
+        with warnings.catch_warnings():
+            warnings.simplefilter('error', NumbaWarning)
+
+            f = mod.closure1
+            self.assertPreciseEqual(f(3), 6) # 3 + 3 = 6
+            f = mod.closure2
+            self.assertPreciseEqual(f(3), 8) # 3 + 5 = 8
+            f = mod.closure3
+            self.assertPreciseEqual(f(3), 10) # 3 + 7 = 8
```
Comment (stuartarchibald) [inline numba/tests/test_caching.py:435]: 3 + 7 = 10 !

File: numba/tests/test_caching.py [hunk 3]
```diff
@@ -133,3 +192,886 @@ def test(num=1000):
...
+        # Test caching in an IPython session
+        base_cmd = [sys.executable, '-m', 'IPython']
+        base_cmd += ['--quiet', '--quick', '--no-banner', '--colors=NoColor']
+        try:
+            ver = subprocess.check_output(base_cmd + ['--version'])
+        except subprocess.CalledProcessError as e:
+            self.skipTest("ipython not available: return code %d"
+                          % e.returncode)
+        ver = ver.strip().decode()
+        print("ipython version:", ver)
```
Comment (stuartarchibald) [inline numba/tests/test_caching.py:605]: delete?

File: numba/tests/test_dispatcher.py (hunk 1 above)
Comment (gmarkall) [reply numba/tests/test_dispatcher.py:1090]: It is.

**Comment by gmarkall** [comment]
@stuartarchibald Many thanks for the review, your comments should be addressed now.

File: numba/tests/test_dispatcher.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/tests/test_dispatcher.py:1090]: Thanks, its removal makes sense!

**Comment by esc** [comment]
Thank you for the patch!

**Pull request merged**